\documentclass[10pt]{article}
\usepackage{color}
\definecolor{gray}{rgb}{0.7,0.7,0.7}
\usepackage{framed}
\usepackage{enumitem}
\usepackage{longtable}
\usepackage[pdfborder={0 0 0},hyperfootnotes=false]{hyperref}

\addtolength{\textwidth}{3.4cm}
\addtolength{\hoffset}{-1.7cm}
\addtolength{\textheight}{4cm}
\addtolength{\voffset}{-2cm}


\begin{document}

\title{metadamage(?) formats}
\author{tsk}
\maketitle
\vspace*{1em}

This document describe some of the internal formats used by the metadamage software. These are at the current time.
\begin{itemize}
\item .bdamage.gz
\item .lca
\item .stat
\end{itemize}

\section{bdamage format}
bdamage files are files that contain counts of mismatchs conditional on strand and cycle (position within read). These are generated with metadamage lca or metadamage getdamage. 
The first 8 bytes magic number determines which bdamage version. If no magic number is present then version0 is assumed. A magic number of \emph{bdmgv1} means version1.

\subsection{version 0}
First version of the bdamage file is a single bgzf compressed file. MAXLENGTH occurs once in the beginning of the file, followed by succesive blocks of data[1-8]. Block[3-5] indicates the actual mismatch counts for the forward which we have MAXLENGTH times. Block[6-8] indicates the actual mismatch counts for the reverse strand which will also occur MAXLENGTH times. 
\begin{table}[h!]
\begin{tabular}{rllll}
  \hline
  {\bf Col} & {\bf Field} & {\bf Type} & {\bf Brief description} \\
  \hline
  0 & {\sf MAXLENGTH} & int &  Number of cycles\\\hline
  1 & {\sf ID} & int &  Id for mismatch type$^1$\\
  2 & {\sf NREADS} & int & Number of reads used supporting the mismatch matrix\\\hline\hline
  3 & {\sf 1} & int[16] & mismatch rate for first cycle from the 5prime\\
  4 & {\sf $i$} & int[16] & mismatch rate for the $i$'th cycle from the 5prime\\
  5 & {\sf MAXLENGTH} & int[16] & mismatch rate for the last cycle from the 5prime\\\hline
  6 & {\sf 1} & int[16] & mismatch rate for first cycle from the 3prime\\
  7 & {\sf $i$} & int[16] & mismatch rate for the $i$'th cycle from the 3prime\\
  8 & {\sf MAXLENGTH} & int[16] & mismatch rate for the last cycle from the 3prime\\\hline
  \hline
\end{tabular}\label{tab1}
\caption{Content of bdamage.gz file. Note$^1$ This is either the taxidID or the referenceID relative to the SAM/BAM header for single species resequencing projects or it is the \textit{taxid} if output has been generated with metadamage lca 3) Order is given by AA,AC,AG,AT,CA,CC,CG,CT,GA,GC,GG,GT,TA,TC,TG,TT, with first base indicatting reference nucleotide and second base indicating observed nucleotide}.
\end{table}

\subsection{version 1}
First version of the bdamage file is a single bgzf compressed file. MAXLENGTH occurs once in the beginning of the file, followed by succesive blocks of data[1-8]. Block[3-5] indicates the actual mismatch counts for the forward which we have MAXLENGTH times. Block[6-8] indicates the actual mismatch counts for the reverse strand which will also occur MAXLENGTH times. 
\begin{table}[h!]
\begin{tabular}{rllll}
  \hline
  {\bf Col} & {\bf Field} & {\bf Type} & {\bf Brief description} \\
  \hline
  0 & {\sf MAXLENGTH} & int &  Number of cycles\\\hline
  1 & {\sf ID} & int &  Id for mismatch type$^1$\\
  2 & {\sf NREADS} & int & Number of reads used supporting the mismatch matrix\\\hline\hline
  3 & {\sf NREADS\_PER\_SPECIES} & int & Number of unique reads per species\\\hline\hline
  4 & {\sf 1} & int[16] & mismatch rate for first cycle from the 5prime\\
  5 & {\sf $i$} & int[16] & mismatch rate for the $i$'th cycle from the 5prime\\
  6 & {\sf MAXLENGTH} & int[16] & mismatch rate for the last cycle from the 5prime\\\hline
  7 & {\sf 1} & int[16] & mismatch rate for first cycle from the 3prime\\
  8 & {\sf $i$} & int[16] & mismatch rate for the $i$'th cycle from the 3prime\\
  9 & {\sf MAXLENGTH} & int[16] & mismatch rate for the last cycle from the 3prime\\\hline
  \hline
\end{tabular}\label{tab1}
\caption{Content of bdamage.gz file. Note$^1$ This is either the taxidID or the referenceID relative to the SAM/BAM header for single species resequencing projects or it is the \textit{taxid} if output has been generated with metadamage lca 3) Order is given by AA,AC,AG,AT,CA,CC,CG,CT,GA,GC,GG,GT,TA,TC,TG,TT, with first base indicatting reference nucleotide and second base indicating observed nucleotide}.
\end{table}


\clearpage
\section{.lca}
This section describes the test output generated by a metadamage lca subfunctionality and contains information at the readlevel regarding both taxonomic information and statistics pertaining usefull readinformation.\\

First line of the file begins with a hashtag followed by the actual command used for generating the file. Last entry of the line is again a hashtag followed by the git commit id which will serve as a primitive versioncontrol.

Each line consists of a number of items seperated by tabspace. First entry contains readID together with other information seperated by colon. After the first entry succesive blocks of the type taxid:name:''taxlevel'' from the lca toward the root. The complete specification is seen in table below.

\begin{table}[h]
\begin{tabular}{rll}
  \hline
  {\bf Col} & {\bf Brief description} \\
  \hline
  1 & {\sf readID} & readID, this might contains colon\\
  2 & {\sf seq} & The actual sequence\\
  3 & {\sf length(seq)} & The length of the sequence\\
  4 & {\sf nAlignments} & The number of alignments used for inferring the lca\\
  5 & {\sf gc-content} & The GC content for the sequence\\\hline\hline
  6 & {\sf taxid} & the taxomic id (integer)\\
  7 & {\sf taxid} & the taxonomic name(string)\\
  8 & {\sf ``taxlevel''} & the taxonomic level\\\hline\hline      
\end{tabular}\label{tab2}
\caption{Content of a .lca file. Note that 1) seperate between field[1-5] is colon, but the readID might also contain colon. 2) the quotes around field[8] is intentional since taxlevels might contain spaces. 3) Not that the number of tab seperated entries is different between reads since this is the number of nodes needed to traverse from lca to root}
\end{table}
\section{.stat}
Very simple tabseperated flatfile
\begin{enumerate}
\item taxid
\item Number of supporting reads
\item Mean lengths of supporting reads
\item Variance of the lengths of the supporting reads
\item Mean gccontent of supporting reads
\item Variance of the gccontent of supporting reads
\item name of lca in quotes (if relevant, otherwise NA)
\item name of taxomic level of lca in quotes (if relevant, otherwise NA)
\end{enumerate}
\end{document}

%%% Local Variables:
%%% mode: latex
%%% TeX-master: t
%%% End:
